### ISSUE TITLE
Incorrect link counts in Linux on Asus ProArt X570

### ISSUE THREAD

**Issue by mwpastore** [open-mpi/hwloc#528]
### What version of hwloc are you using?

```
mwp@hyperion:~$ lstopo --version
lstopo 2.4.1
mwp@hyperion:~$ dpkg -l '*hwloc*'
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name                     Version      Architecture Description
+++-========================-============-============-==============================================
ii  hwloc                    2.4.1+dfsg-1 amd64        Hierarchical view of the machine - utilities
un  libhwloc-contrib-plugins <none>       <none>       (no description available)
ii  libhwloc-plugins:amd64   2.4.1+dfsg-1 amd64        Hierarchical view of the machine - plugins
ii  libhwloc15:amd64         2.4.1+dfsg-1 amd64        Hierarchical view of the machine - shared libs
```

### Which operating system and hardware are you running on?

```
mwp@hyperion:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 21.10
Release:	21.10
Codename:	impish
mwp@hyperion:~$ uname -a
Linux hyperion 5.16.0-18.1-liquorix-amd64 #1 ZEN SMP PREEMPT liquorix 5.16-26ubuntu1~impish (2022-03-28) x86_64 x86_64 x86_64 GNU/Linux
```

#### Asus ProArt X570-CREATOR WiFi

Ryzen 5000 on this platform presents four (4) dedicated Gen4 lanes for the chipset and four (4) dedicated Gen4 lanes for the primary M.2 slot, and supports x8x4x4 bifurcation of the 16 Gen4 PEG lanes. The following PCIe devices are connected via CPU lanes: 

##### Primary M.2

* M.2_1 (01:00.0):  2TB PCIe NVMe SSD (Gen3 x4)

##### PEG

* PCIEX16_1 (3e:00.0): NVIDIA RTX 3080 12GB (Gen4 x8, capable of x16)
* PCIEX16_2 (3f:00.0): AMD Radeon Pro WX 3200 (Gen3 x4, capable of x8)
* M.2_2 (40:00.0): 1TB PCIe NVMe SSD (Gen3 x4)

<img width="875" alt="image" src="https://user-images.githubusercontent.com/8540936/161321395-3f520c8d-0004-4602-93e2-d0aed0773c8d.png">

### Details of the problem

* shows a x8 connection to the chipset, this should be x4
* shows a x16 connection to PCIEX16_1 (3e:00.0), this should be x8
* shows a x1 connection to PCIEX16_2 (3f:00.0), this should be x4

Relevant snippets from `dmesg`:

```
[    0.583659] pci 0000:3e:00.0: 126.024 Gb/s available PCIe bandwidth, limited by 16.0 GT/s PCIe x8 link at 0000:00:03.1 (capable of 252.048 Gb/s with 16.0 GT/s PCIe x16 link)
[    0.584084] pci 0000:3f:00.0: 31.504 Gb/s available PCIe bandwidth, limited by 8.0 GT/s PCIe x4 link at 0000:00:03.2 (capable of 63.008 Gb/s with 8.0 GT/s PCIe x8 link)
```

### Additional information

[Archive.zip](https://github.com/open-mpi/hwloc/files/8400182/Archive.zip)

**Comment by mwpastore** [comment]
I have a different Asus X570 board with a Ryzen 3000 processor, less complex bifurcation, and very similar issues if you'd be interested in seeing the details on that. It's showing the x4 primary M.2 as x8, the x4 chipset as x8, and x8x8 PEG as x2x16.

**Comment by bgoglin** [comment]
I would need the output of hwloc-gather-topology with --io or a verbose lspci to debug this, but there's nothing wrong on the hwloc side since we're just reading what Linux reports in sysfs. You should run "sudo lspci -vv" and look for LnkCap (what the link support) and LnkSta (what's currently selected). If it's different from dmesg, you should ask kernel people, but usually it just means that a driver reduced the link speed. At least the NVIDIA driver does that the GPU is idle. So the lstopo/lspci outputs vary depending on what's running on the machine.

Anyway, what you see in lstopo is link bandwidth in GB/s, not the number of lanes. So 16.0 on the left of 3e:00.0 means 16GB/s, which is correct for 8x lanes of PCIe Gen4. 1.0 on the left of 3f:00.0 means 1GB/s which is indeed less than 4GB/s for 4x lanes of PCIe Gen3. Maybe the AMD driver decreases the speed there too.
I am not sure where you see a 8x in the output.

**Comment by mwpastore** [comment]
@bgoglin Oh my gosh, how embarrassing. I've been looking at `lstopo` for years and never realized it was bandwidth, not number of lanes. Thank you for kindly showing me the light. 

I have seen both GPUs step down in GPU-Z as you've described. 3e:00.0 is currently active and 3f:00.0 is currently idle, so that checks out. 

The "x8" I saw is actually the 7.9 link that has multiple descendants. This is the chipset link. Applying what I've just learned, this matches the throughput of a Gen4 x4 link, which is correct for this platform. 

Given this new context, the various `lstopo` outputs I've been puzzling over now appear to be absolutely, 100% correct. Thank you!
